src/Storages/MergeTree/ReplicatedMergeMutateTaskBase.h

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include <common/logger_useful.h>
+
+#include <Storages/MergeTree/IExecutableTask.h>
+#include <Storages/MergeTree/ReplicatedMergeTreeQueue.h>
+
+namespace DB
+{
+
+class StorageReplicatedMergeTree;
+
+/**
+ * This is used as a base of MergeFromLogEntryTask and MutateFromLogEntryTaskBase
+ */
+class ReplicatedMergeMutateTaskBase : public IExecutableTask
+{
+public:
+    template <class Callback>
+    ReplicatedMergeMutateTaskBase(
+        Poco::Logger * log_,
+        StorageReplicatedMergeTree & storage_,
+        ReplicatedMergeTreeQueue::SelectedEntryPtr & selected_entry_,
+        Callback && task_result_callback_)
+        : selected_entry(selected_entry_)
+        , entry(*selected_entry->log_entry)
+        , log(log_)
+        , storage(storage_)
+        , task_result_callback(task_result_callback_) {}
+
+    ~ReplicatedMergeMutateTaskBase() override = default;
+
+    void onCompleted() override;
+
+    StorageID getStorageID() override;
+
+    bool executeStep() override;
+
+protected:
+    void prepareCommon();
+    virtual bool prepare() = 0;
+    bool executeImpl() ;
+    virtual bool finalize() = 0;
```

Comment:
I think that not all of this should be protected.
`prepare` and `finalize` probably are, `prepareCommon` and `executeImpl` probably not.
Also, `state` and `task_result_callback` are probably provate

Reply:
Add a comment what is the semantic for `bool` return value.

Reply:
> Also, state and task_result_callback are probably provate

There is a non-consistency, because the state could be used in prepare method. When we understand that we don't have to execute a task and it is Ok.

Reply:
Then return special state from prepare. It is more difficult if state can be changed somewhere else

Reply:
Fixed.